WordPress/gutenberg
# Call variation through callback so it's only loaded when needed - in support of  trac 59969

## What?
This PR will help build variations only when we need to load them instead of during resignation at the `init` hook. 

Core PR with detailed performance analysis -> 
https://github.com/WordPress/wordpress-develop/pull/5718

## Why?
Currently, we run a performance-heavy build variation function for some of the blocks in `init` hook, which causes it to load even on the frontend where it is not needed.

## How?
In this PR we register only the callback during block registration and build the variation only when they are needed.

**Note: This PR provides around ~9% improvement to the frontend load time but it has no improvement for the editor site. However, we have found some way by which similar performance benefits could be obtained at the editor side -  https://github.com/WordPress/wordpress-develop/pull/5718#issuecomment-1845361155**


## Testing Instructions
Test it for the TT4 theme. 
1. Open the editor. 
2. Open the browser console and type `wp.blocks.getBlockType('core/post-terms')`
3. Ensure that the block object returned has variation. Do the same for `core/navigation-link`, `core/template-part`  

### Testing Instructions for Keyboard


## Screenshots or screencast
